(compat) Add documentation for cross-client compatibility (#26379)
microsoft/FluidFramework: CrossClientCompatibility.md

## Overview

Cross-client compatibility is Fluid's ability to support collaboration between two clients running different versions of Fluid within an allowable window. This addresses two key scenarios:
1. **Rolling upgrades**: During version upgrades, there is an unavoidable transition window when clients running different versions must coexist and collaborate. This compatibility ensures users can continue working together seamlessly, whether or not their application instance has been updated yet.
2. **Multi-application ecosystems**: Different applications with different deployment schedules may host the same Fluid content. In such ecosystems, all applications integrating Fluid-based experiences must coordinate to respect the cross-client compatibility window. This avoids requiring all applications to be on exactly the same version, which would be impractical.

> > **Note:** The cross-client compatibility guarantee applies to all Fluid layers (Driver, Loader, Runtime, and Datastore). However, the enforcement mechanisms described in this document — `minVersionForCollab`, feature gating, and client version checks — are currently implemented only at the **Runtime and Datastore layers**. The Driver layer does not currently have cross-client compatibility concerns because it does not exchange data formats between clients. Enforcement at the Loader layer may be added in the future. See the [Interaction with Layer Compatibility](./FluidCompatibilityConsiderations.md#interaction-with-layer-compatibility) section for more details.

Reviewer: I am wondering if we should add a note that this is only relevant if using encapsulated model (since this doc talks about that later) and point it to the section that talks about it.

Author: I think it is mostly still relevant for declarative model since they both use the same underlying enforcement mechanisms, but I added a note that says configuration depends on your application model and linked to that section.

> **Note:** When referring to previous major releases (e.g., N-1), we assume the latest minor release within that major version series. All minor releases within a given major version are compatible with each other. However, please note this is currently TBD and is likely subject to change.

## Cross-Client Compatibility Policy

The Fluid Framework guarantees cross-client compatibility between adjacent major versions. This policy ensures that applications have sufficient time to upgrade while maintaining seamless collaboration.

| Version Combination | Compatibility |
|---------------------|---------------|
| **N / N-1** (adjacent major versions) | ✅ Fully compatible. No special action required. |
| **N / N-2 or older** | ❌ Not supported. These clients cannot collaborate. |

**Enforcement:** Incompatible clients will be blocked from collaborating on a document and shown a clear error message (see [Errors and Warnings to Monitor](#errors-and-warnings-to-monitor)).

**Example:** If the most recent public major release (N) is 4.0, a client running 4.x is cross-client compatible with 3.x clients, but not with 2.x or older clients.

> ## Understanding `minVersionForCollab`

Reviewer: I would suggest adding a heading `## Cross-client Compability Configuration and Enforcment` (or something similar) and make this a subheading `### minVersionForCollab` (no `` around minVersionForCollab). The sections through to `## Best Practices` should be under the Config and Enforcement heading.

# Fluid Framework Cross-Client Compatibility

**Suggested pre-read:** The [Fluid Framework Compatibility Considerations](./FluidCompatibilityConsiderations.md) document provides an overview of Fluid's four compatibility dimensions: API compatibility, Layer compatibility, Cross-client compatibility, and Data-at-rest compatibility. This document focuses specifically on Cross-client compatibility — how clients running different versions of the Fluid runtime can collaborate on the same document.

## Overview

Cross-client compatibility is Fluid's ability to support collaboration between two clients running different versions of the Fluid runtime within an allowable window. This addresses two key scenarios:
1. **Rolling upgrades**: During version upgrades, there is an unavoidable transition window when clients running different versions must coexist and collaborate. This compatibility ensures users can continue working together seamlessly, whether or not their application instance has been updated yet.

> 1. **Multi-application ecosystems**: Different applications with different deployment schedules may host the same Fluid content. In such ecosystems, all applications integrating Fluid-based experiences must coordinate to respect the cross-client compatibility window. This avoids requiring all applications to be on exactly the same version, which would be impractical.

Reviewer: These numbers are all `1` here and in other places

This document explains:

- Which versions of the Fluid runtime can collaborate together
- How to configure cross-client compatibility
- Best practices for upgrading safely
- What to monitor in telemetry

### Terminology

| Term | Definition |
|------|------------|
| **N** | The most recent public major release of the Fluid Framework |
| **N-1** | The second most recent public major release |
| **N-2** | The third most recent public major release |
| **Saturation** | When an adequate percentage of an application's clients are running a certain version. The threshold that is considered adequate is defined by the application's requirements.  |

> > **Note:** When referring to previous major releases (e.g., N-1), we assume the latest minor release within that major version series. All minor releases within a given major version are compatible with each other.

Author: Yes, this will probably change and figuring this out is the next compat item I plan to work on. I figured for now we can put what we had in the previous policy doc and update it when it changes.

## Cross-Client Compatibility Policy

The Fluid Framework guarantees cross-client compatibility between adjacent major versions. This policy ensures that applications have sufficient time to upgrade while maintaining seamless collaboration.

| Version Combination | Compatibility |
|---------------------|---------------|

> | **N / N-1** (adjacent major versions) | ✅ Fully compatible by default. No special action required. |

Reviewer: ```suggestion
| **N / N-1** (adjacent major versions) | ✅ Fully compatible. No special action required. |
```
I don't think we need to specify `by default` here since there are no alternatives.

| **N / N-2 or older** | ❌ Not supported. These clients cannot collaborate. |

> **Enforcement:** Incompatible clients will be blocked from collaborating on a document and shown a clear error message.

Reviewer: You should link the section that describes the error. Maybe consider having that section before the policy and then link to it.

**Example:** If the most recent public major release (N) is 4.0, a client running 4.0 is cross-client compatible with 3.x clients by default, but not with 2.x or older clients.

## What This Means for An Application

As an application developer, you need to manage your Fluid version upgrades carefully to ensure uninterrupted collaboration for your users. By configuring `minVersionForCollab` appropriately and monitoring your client version distribution, you can safely upgrade while maintaining compatibility across your user base.

### Configuring Cross-Client Compatibility

> For the best results, we encourage you to use the "proper configurations" for cross-client compatibility. In this case, the proper configurations would be setting the `minVersionForCollab` property (defined in [containerRuntime.ts](./packages/runtime/container-runtime/src/containerRuntime.ts)). `minVersionForCollab` represents the minimum Fluid runtime version allowed to collaborate in a document. It will automatically disable any features that prevent safe cross-client collaboration for clients that are at least the version specified.

Reviewer: ```suggestion
For the best results, we encourage you to set the `minVersionForCollab` property (defined in [containerRuntime.ts](./packages/runtime/container-runtime/src/containerRuntime.ts)). `minVersionForCollab` represents the minimum Fluid runtime version allowed to collaborate in a document. It will automatically disable any features that prevent safe cross-client collaboration for clients that are at least the version specified.
```

I don't think the first sentence there is needed.

1. Monitor telemetry for warnings/errors to ensure safe rollout (see below section). At this point any clients that are not saturated may be blocked from accessing the document.

### Errors and Warnings to Monitor

The following are errors and telemetry warnings you may see during and following an upgrade. Monitoring these signals will help ensure a safe rollout.

| Signal | What it Means | What to Do |
|--------|---------------|------------|
| Telemetry Event: `MinVersionForCollabWarning` | Clients are joining with a version below your configured minimum, but are still able to understand the document's data format and therefore continue to collaborate. | If you see this warning message, it's likely a sign you updated `minVersionForCollab` too quickly. In future releases, ensure proper saturation before updating. If these warning messages are ignored, you may risk seeing the below error in the future. |

> | `DataProcessingError`: Document can't be opened with current version of the code | An out-of-window client tried to join and was blocked due to being unable to collaborate with the newer client's document. | If this was unexpected, lower `minVersionForCollab` to allow older clients to join. |

Reviewer: Maybe give examples?
